### DISCUSSION TITLE
Next.config.js not loaded in Storybook (and probably in other 3rd party tools)

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/next.js#11143] [Help]
# Bug report

## Describe the bug

Next.config.js `env` and runtime configs are not taken into account in Storybook stories. It makes code dependant on `next/config` difficult to test.

## To Reproduce

Full repro here: https://github.com/eric-burel/with-storybook-config-repro

- init storybook example
```sh
npm init next-app --example with-storybook with-storybook-app
```
- add `next.config.js` at root
```js
module.exports = (nextConfig = {}) => {
  return {
    env: { "foo":"bar"},
    publicRuntimeConfig: { "hello":"world"}
 };
};
```
- add following code in the demo story, eg welcome
```js
import getConfig from "next/config"
console.log("getConfig", getConfig())
console.log("process.env.foo", process.env.foo)
```
- run storybook
```
yarn run storybook
```
- see browser console, variables are undefined

## Expected behavior

`next.config.js` should be somehow taken into account, at least for the relevant part.

## System information

- Version of Next.js: 0.3

## Additional context

I work with TypeScript and I've tried to add a `.babelrc` with next/babel config.

**Comment by eric-burel** [comment]
Related to current RFC about env https://github.com/zeit/next.js/discussions/11106

**Comment by eric-burel** [comment]
Reading the code, it seems that the `loadConfig` function find the file, but I am not sure which part actually make it available to `getConfig` or handles `env`.

**Comment by eric-burel** [comment]
Current quickfix: I use webpack alias to mock.

```.storybook/mocks/next-config.js
// TODO: use `loadConfig` from `packages/next/next-server/server/config.ts` to load the dev next config
module.exports = function getConfig() {
  return {
    publicRuntimeConfig: {},
    serverRuntimeConfig: {}
  };
};
```
In my Storybook Webpack config:
```js
    config.resolve.alias = {
      ...(config.resolve.alias || {}),
     // magic to get nicer imports
      "~": path.join(__dirname, "../src"),
      // example loading a mock package
      "next/config": path.join(__dirname, "./mocks/next-config.js")
    };
```

That's for runtime config. 
For build config, replacing `process.env` is trickier, that would require a Webpack loader.

**Reply by Ivaylo-Korakov** [reply]
Where does this `path` variable come from?

**Reply by eric-burel** [reply]
Jsut the `path` core library: `import path from "path"`.
Btw nowadays I wouldn't recommend this anymore, you should not use `next/config`. Basically, just create a JS file and load it.
See my article on the topic for more advanced patterns: https://blog.vulcanjs.org/how-to-set-configuration-variables-in-next-js-a81505e43dad

**Comment by rafaelalmeidatk** [comment]
This is not really a bug, it is expected since next.config.js is only utilized by Next. Storybook has its own webpack config, the same for Jest. There are some stuff that you can still share like the babel config. I also had this "problem" with local aliases and solved it by using `babel-plugin-module-resolver`, since it is defined with babel both libs will use it.

**Comment by eric-burel** [comment]
Yeah not really a bug, more a feature request/a design question, but in the end we need a pattern to handle this. Initially I thought that the `next` preset would handle that kind of things but of course that's more complicated. Maybe some internals of Next could be exposed to make it easier for 3rd party to load a next.config.

`babel-plugin-module-resolver` seems like a great idea too.

**Comment by eric-burel** [comment]
Nowadays every "build feature" will end up being necessary to another 3rd party tool that must work with the code, be it Storybook or a test framework etc. so it's good to expose them as much as possible for integration, or at least document how to mock them.
